Minor deserialisation optimisations

SquidDev opened the pull request:
Related to #5977, I've been fiddling with a couple of other minor optimisations to the deserialiser:

 - Use an `Array Int32 (HashMap TypeRepr U)` for caching deserialised nodes, rather than `HashTable (Int32, TypeRepr) U`

   The intuition here is that _most_ nodes will only deserialise to a single type type, and so its quicker to do an array lookup + lookup in a singleton HashMap (the latter of which is cheap) compare with a lookup in a hashtable (which requires several array accesses).

 - Deserialise "straight" to an array:

   The current `decode` implementation reads/writes a list of a type (strings, numbers, nodes) using `Binary.get`, gets the length of that array, and then converts it to an array.

   Replacing the call to `genericLength :: [a] -> Int32` with `fromIntegral . genericLength` significantly brings a reasonable boost (0.6s quicker in my test project). However, we can avoid computing the length entirely by defining our own `Binary` instance which can reuse the length, gaining another 0.3s.

   It's possible writing our own `Binary` instance specialises better too - I've tried to dig into the core, but it's a very noisy diff to wade through.

As in #5977, benchmarks were taken on a non-profiling build built with GHC 8.10.7, then tested by loading [this project](https://github.com/plt-amy/1lab). I've not tested this on other projects, compilers or machines, so not sure how consistent these improvements will be (somewhat scared to test it on the whole stdlib, given how long it takes to TC, but not sure if there's other good projects to benchmark?).

Definitely not an expert on any of this, and not sure how sensible this patch is, so happy to make any changes!

<details><summary>Baseline (#5977)</summary>

```
Total                      9,508ms
Miscellaneous                  4ms
Deserialization            7,103ms (9,412ms)
Deserialization.Compaction 2,308ms
Import                        69ms
Parsing                       21ms
  10,808,268,736 bytes allocated in the heap
   3,967,777,072 bytes copied during GC
     285,751,936 bytes maximum residency (44 sample(s))
       4,242,320 bytes maximum slop
             515 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      8977 colls,     0 par    3.248s   3.270s     0.0004s    0.0173s
  Gen  1        44 colls,     0 par    0.547s   0.548s     0.0124s    0.0474s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    5.729s  (  5.693s elapsed)
  GC      time    3.796s  (  3.818s elapsed)
  EXIT    time    0.001s  (  0.009s elapsed)
  Total   time    9.527s  (  9.521s elapsed)

  Alloc rate    1,886,509,235 bytes per MUT second

  Productivity  60.1% of total user, 59.8% of total elapsed
```

</details>

<details><summary>Use an array for memoising deserialised values (783ae634f1f754b94a74568c105293a1297bbe3c)</summary>

```
Total                      8,464ms
Miscellaneous                  3ms
Deserialization            5,805ms (8,385ms)
Deserialization.Compaction 2,579ms
Import                        61ms
Parsing                       14ms
   8,407,722,464 bytes allocated in the heap
   3,794,800,792 bytes copied during GC
     278,672,144 bytes maximum residency (36 sample(s))
         233,256 bytes maximum slop
             528 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      7127 colls,     0 par    3.440s   3.462s     0.0005s    0.0177s
  Gen  1        36 colls,     0 par    0.434s   0.434s     0.0121s    0.0477s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    4.615s  (  4.589s elapsed)
  GC      time    3.874s  (  3.896s elapsed)
  EXIT    time    0.001s  (  0.005s elapsed)
  Total   time    8.491s  (  8.490s elapsed)

  Alloc rate    1,821,966,522 bytes per MUT second

  Productivity  54.3% of total user, 54.0% of total elapsed
```

</details>

<details><summary>Deserialise directly to an array (73dcce0e76b262daff2a0d1f293b0ed25abd1da5)</summary>

```
Total                      7,524ms          
Miscellaneous                  6ms          
Deserialization            4,987ms (7,415ms)
Deserialization.Compaction 2,428ms          
Import                        62ms          
Parsing                       29ms          
ModuleName                    10ms          
   7,573,159,984 bytes allocated in the heap
   3,359,548,336 bytes copied during GC
     275,288,952 bytes maximum residency (33 sample(s))
         314,592 bytes maximum slop
             486 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0      6510 colls,     0 par    2.909s   2.925s     0.0004s    0.0180s
  Gen  1        33 colls,     0 par    0.355s   0.356s     0.0108s    0.0481s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time    4.275s  (  4.243s elapsed)
  GC      time    3.265s  (  3.281s elapsed)
  EXIT    time    0.001s  (  0.006s elapsed)
  Total   time    7.542s  (  7.531s elapsed)

  Alloc rate    1,771,375,687 bytes per MUT second

  Productivity  56.7% of total user, 56.3% of total elapsed
```

</details>

SquidDev committed:
Use an array for memoising deserialised values

We now use an array of hashmaps, rather than a single hashmap of
(idx, type). As most slots will be a single entry, this makes lookups
faster, resulting in a 10% speedup.

SquidDev committed:
Deserialise directly to an array

Not actually, as Binary doesn't allow reading to an array. However, this
saves an extra call to List.genericLength and appears to specialise
better, saving another 10%.

JacquesCarette commented:
You should try to load up 1) stdlib, 2) agda-categories. 

Also: nice work!

SquidDev commented:
Ahh good, they appear to be pretty consistent with my test project, ~20% reduction in runtime and 30% reduction in allocations.

## agda-categories

<details><summary>Before</summary>

```
Total                      22,307ms           
Miscellaneous                   7ms           
Deserialization            16,142ms (22,096ms)
Deserialization.Compaction  5,954ms           
Import                        167ms           
Parsing                        35ms           
  31,293,224,696 bytes allocated in the heap
   1,425,371,384 bytes copied during GC
     728,583,720 bytes maximum residency (9 sample(s))
       5,183,144 bytes maximum slop
             995 MiB total memory in use (37 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       192 colls,     0 par    2.188s   2.189s     0.0114s    0.0644s
  Gen  1         9 colls,     0 par    0.192s   0.192s     0.0214s    0.0398s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.002s  (  0.002s elapsed)
  MUT     time   19.948s  ( 19.905s elapsed)
  GC      time    2.380s  (  2.381s elapsed)
  EXIT    time    0.001s  (  0.002s elapsed)
  Total   time   22.331s  ( 22.291s elapsed)

  Alloc rate    1,568,750,264 bytes per MUT second

  Productivity  89.3% of total user, 89.3% of total elapsed
```

</details>


<details><summary>After</summary>

```
Total                      17,503ms           
Miscellaneous                   7ms           
Deserialization            11,342ms (17,282ms)
Deserialization.Compaction  5,940ms           
Import                        180ms           
Parsing                        34ms           
  22,267,561,176 bytes allocated in the heap
     907,400,296 bytes copied during GC
     633,637,376 bytes maximum residency (8 sample(s))
       1,923,000 bytes maximum slop
             989 MiB total memory in use (30 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       143 colls,     0 par    1.297s   1.299s     0.0091s    0.0691s
  Gen  1         8 colls,     0 par    0.223s   0.223s     0.0278s    0.0950s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.003s  (  0.003s elapsed)
  MUT     time   16.006s  ( 15.974s elapsed)
  GC      time    1.520s  (  1.521s elapsed)
  EXIT    time    0.001s  (  0.002s elapsed)
  Total   time   17.530s  ( 17.501s elapsed)

  Alloc rate    1,391,187,863 bytes per MUT second

  Productivity  91.3% of total user, 91.3% of total elapsed
```

</details>

## agda-stdlib

<details><summary>Before</summary>

```
Total                      42,899ms           
Miscellaneous                   3ms           
Deserialization            31,222ms (42,541ms)
Deserialization.Compaction 11,319ms           
Import                        303ms           
Parsing                        51ms           
  46,486,510,992 bytes allocated in the heap
  17,044,516,784 bytes copied during GC
   1,176,535,616 bytes maximum residency (51 sample(s))
       5,318,520 bytes maximum slop
            1830 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     39002 colls,     0 par   16.128s  16.220s     0.0004s    0.0296s
  Gen  1        51 colls,     0 par    1.613s   1.613s     0.0316s    0.1520s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time   25.224s  ( 24.985s elapsed)
  GC      time   17.741s  ( 17.833s elapsed)
  EXIT    time    0.001s  (  0.002s elapsed)
  Total   time   42.967s  ( 42.821s elapsed)

  Alloc rate    1,842,931,086 bytes per MUT second

  Productivity  58.7% of total user, 58.3% of total elapsed
```

</details>

<details><summary>After</summary>

```
Total                      33,632ms           
Miscellaneous                   4ms           
Deserialization            21,467ms (33,272ms)
Deserialization.Compaction 11,804ms           
Import                        303ms           
Parsing                        51ms           
  32,974,676,424 bytes allocated in the heap
  14,593,547,232 bytes copied during GC
   1,196,561,120 bytes maximum residency (39 sample(s))
         728,000 bytes maximum slop
            1813 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     28671 colls,     0 par   13.841s  13.911s     0.0005s    0.0278s
  Gen  1        39 colls,     0 par    1.048s   1.049s     0.0269s    0.1380s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time   18.806s  ( 18.639s elapsed)
  GC      time   14.889s  ( 14.960s elapsed)
  EXIT    time    0.001s  (  0.001s elapsed)
  Total   time   33.697s  ( 33.601s elapsed)

  Alloc rate    1,753,397,524 bytes per MUT second

  Productivity  55.8% of total user, 55.5% of total elapsed
```

</details>

JacquesCarette commented:
I think that makes it quite convincing that it's not a fluke, this is a worthwhile improvement.

SquidDev commented:
I'm able to eck out another 5% by writing to an array directly, rather than going through a list (bringing the stdlib down to 31.4s and agda-categories to 16.4s). However I make liberal use of `unsafePerformIO` - I don't know if this is a change people would be comfortable integrating? I know for one I'm never confident writing code which relies on it!

```haskell
instance B.Binary a => B.Binary (ListLike a) where
  put = __IMPOSSIBLE__ -- Will never deserialise this
  get = do
    n <- B.get :: B.Get Int
    arr <- return (unsafePerformIO $ newArray_ (0, fromIntegral n - 1) :: IOArray Int32 a)

    let
      getMany i = if i == n then return () else do
        x <- B.get
        () <- return $ unsafePerformIO $ unsafeWrite arr i x
        getMany (i + 1)
    () <- getMany 0

    arr <- return $ unsafePerformIO $ unsafeFreeze arr
    return $! ListLike arr
```

andreasabel commented:
Great work, @SquidDev !

Concerning `unsafePerformIO`,  I share your confidence issues.  Alternatives:
- Could one instead replace the `Binary instance` for `Listlike` by some method living in `IO` and use it as such from `decode` (which lives in `TCM` and has `IO`)?  
- Give precise reasoning why this pattern of using `unsafePerformIO` is safe (also when used concurrently).
- Skip the extra 5% (or is it rather 7%) improvement.

nad commented:
Perhaps one could use the `STT` monad transformer (see [Control.Monad.ST.Trans](https://hackage.haskell.org/package/STMonadTrans-0.4.6/docs/Control-Monad-ST-Trans.html)).

SquidDev commented:
>  Could one instead replace the `Binary instance` for `Listlike` by some method living in `IO` and use it as such from `decode` (which lives in `TCM` and has `IO`)?

I think the main problem here is that we'd end up needing to implement most of `Binary` to do this, which is obviously unfortunate.

> Give precise reasoning why this pattern of using `unsafePerformIO` is safe (also when used concurrently).

Looking at the optimised core, I can kinda convince myself. But that's obviously only for one version of GHC. I think the problem is it breaks so many rules of "how to use unsafePerformIO" it's hard to guarantee anything.

Hah, I think I've rather talked myself out of using `unsafePerformIO`.

> Perhaps one could use the `STT` monad transformer (see [Control.Monad.ST.Trans](https://hackage.haskell.org/package/STMonadTrans-0.4.6/docs/Control-Monad-ST-Trans.html)).

Was actually just looking at that! It appears to work, and (unsurprisingly) has the same performance as the unsafe `IO` version, so I think it's worth switching over to that if people are happy to add it as a dependency?

> Skip the extra 5% (or is it rather 7%) improvement.

Should probably clarify here - it _is_ closer to 6%, but results are noisy (±1%, my machine is not set up for benchmarking :)), so erring on the more conservative side.

nad commented:
> Was actually just looking at that! It appears to work, and (unsurprisingly) has the same performance as the unsafe `IO` version, so I think it's worth switching over to that if people are happy to add it as a dependency?

Apparently Andreas is already maintaining this package, so I think it is fine to add it as a dependency.

SquidDev committed:
Deserialise arrays in-place

This uses STT to allocate a mutable array and write to it while
deserialising, instead of allocating a separate array. This provides a
small (5-6%) improvement to deserialisation times.

SquidDev commented:
There we go! Numbers for the curious:

<details><summary>agda-stdlib</summary>

```
Total                      31,345ms           
Miscellaneous                   4ms           
Deserialization            19,621ms (30,964ms)
Deserialization.Compaction 11,342ms           
Import                        301ms           
Parsing                        75ms           
  32,712,768,384 bytes allocated in the heap
  12,955,373,760 bytes copied during GC
   1,217,073,128 bytes maximum residency (39 sample(s))
       1,523,736 bytes maximum slop
            1809 MiB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     28373 colls,     0 par   12.642s  12.768s     0.0004s    0.0271s
  Gen  1        39 colls,     0 par    0.943s   0.943s     0.0242s    0.1166s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time   17.818s  ( 17.591s elapsed)
  GC      time   13.585s  ( 13.711s elapsed)
  EXIT    time    0.002s  (  0.007s elapsed)
  Total   time   31.406s  ( 31.311s elapsed)

  Alloc rate    1,835,899,498 bytes per MUT second

  Productivity  56.7% of total user, 56.2% of total elapsed
```

</details>

<details><summary>agda-categories</summary>

```
Total                      16,236ms           
Miscellaneous                   5ms           
Deserialization            10,179ms (16,039ms)
Deserialization.Compaction  5,860ms           
Import                        174ms           
Parsing                        16ms           
  21,730,347,856 bytes allocated in the heap
     844,509,472 bytes copied during GC
     794,239,640 bytes maximum residency (8 sample(s))
       1,956,608 bytes maximum slop
             990 MiB total memory in use (33 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0       139 colls,     0 par    1.369s   1.371s     0.0099s    0.0562s
  Gen  1         8 colls,     0 par    0.164s   0.164s     0.0205s    0.0537s

  TASKS: 4 (1 bound, 3 peak workers (3 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.003s  (  0.003s elapsed)
  MUT     time   14.726s  ( 14.696s elapsed)
  GC      time    1.534s  (  1.535s elapsed)
  EXIT    time    0.001s  (  0.007s elapsed)
  Total   time   16.264s  ( 16.241s elapsed)

  Alloc rate    1,475,648,977 bytes per MUT second

  Productivity  90.5% of total user, 90.5% of total elapsed
```

</details>

nad commented on src/full/Agda/TypeChecking/Serialise.hs:
```diff
@@ -181,19 +182,18 @@ newtype ListLike a = ListLike { unListLike :: Array Int32 a }
 
 instance B.Binary a => B.Binary (ListLike a) where
   put = __IMPOSSIBLE__ -- Will never deserialise this
```
I guess this comment is incorrect.

SquidDev committed:
Fix wrong serialise/deserialise in comment

andreasabel commented on src/full/Agda/TypeChecking/Serialise.hs:
Rather than using explicit recursion, couldn't `getMany` be written as a `forM [0..n-1] $ \ i -> ...`?

SquidDev replied:
Unfortunately it looks like GHC (at least GHC 8.10.7) doesn't inline/unfold the list construction here, so performance is a bit worse (1-3% slower, 7% more allocations).

Using `forM_ [0..n-1]` is definitely more idiomatic (and shorter!) though. Not sure if you think it's worth it?

andreasabel replied:
If you find that the manual loop `getMany` is more performant, it is ok to leave it there, but please add a comment (basically a digest of our conversation here) that it is there on purpose because `forM_ [0..n-1]` is slower.
(Aside: I wonder whether this is a known GHC issue.)

SquidDev replied:
> I wonder whether this is a known GHC issue

I'm not sure. I've narrowed it down to a smaller test-case, which still fails to unfold on 9.2.3 at least.

Will try to reproduce on one of the beta releases of GHC and see if there's a relevant bug report if it still occurs.

andreasabel reviewed (changes requested):
Thanks for getting rid of unsafePerformIO!

SquidDev committed:
Explain why we're not using for_ here.

andreasabel commented:
Great work, @SquidDev , thanks!
